processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -556,142 +561,79 @@ protected File makeIndexFiles(
       dimConversions.add(Maps.<String, IntBuffer>newHashMap());
     }
 
+    final int TIME_WRITE = 0;
+    final int TIME_COMBINE = 1;
+    final Timer timer = new Timer(2);
+
     for (String dimension : mergedDimensions) {
       final GenericIndexedWriter<String> writer = new GenericIndexedWriter<String>(
           ioPeon, dimension, GenericIndexed.STRING_STRATEGY
       );
       writer.open();
 
-      List<Indexed<String>> dimValueLookups = Lists.newArrayListWithCapacity(indexes.size());
-      DimValueConverter[] converters = new DimValueConverter[indexes.size()];
+      int numMergeIndex = 0;
+      Indexed<String> dimValueLookup = null;
+      Indexed<String>[] dimValueLookups = new Indexed[indexes.size()];
       for (int i = 0; i < indexes.size(); i++) {
         Indexed<String> dimValues = indexes.get(i).getDimValueLookup(dimension);
         if (!isNullColumn(dimValues)) {
-          dimValueLookups.add(dimValues);
-          converters[i] = new DimValueConverter(dimValues);
+          dimValueLookups[i] = dimValueLookup = dimValues;
+          numMergeIndex++;
         }
       }
 
-      Iterable<String> dimensionValues = CombiningIterable.createSplatted(
-          Iterables.transform(
-              dimValueLookups,
-              new Function<Indexed<String>, Iterable<String>>()
-              {
-                @Override
-                public Iterable<String> apply(@Nullable Indexed<String> indexed)
-                {
-                  return Iterables.transform(
-                      indexed,
-                      new Function<String, String>()
-                      {
-                        @Override
-                        public String apply(@Nullable String input)
-                        {
-                          return (input == null) ? "" : input;
```

Comment:
I think NullConverterDimDim stores "" empty str internally but converts empty str back to null when calling getValue, Strings.emptyToNull gets called when doing a lookup by ID, nullToEmpty gets called when doing an add/lookup by value. 

Looks like the null->empty conversion is still happening though in the PeekingIterator of DictionaryMergeIterator